src/v/kafka/client/test/direct_consumer_fixture_test.cc

```diff
@@ -0,0 +1,199 @@
+    }
+
+    topic_assignment make_assignment(
+      const model::topic& topic,
+      std::vector<int> partitions,
+      std::optional<kafka::offset> initial_offset = std::nullopt) {
+        topic_assignment assignment;
+        assignment.topic = topic;
+        for (const auto& p : partitions) {
+            partition_assignment part;
+            part.partition_id = model::partition_id(p);
+            part.next_offset = initial_offset;
+            assignment.partitions.push_back(part);
+        }
+        return assignment;
+    }
+
+    model::node_id get_partition_leader(const model::ntp& ntp) {
+        std::optional<model::node_id> leader_id;
+        model::timeout_clock::time_point deadline = model::timeout_clock::now()
+                                                    + 10s;
+        while (!leader_id) {
+            if (model::timeout_clock::now() > deadline) {
+                throw std::runtime_error(
+                  fmt::format("Timeout while waiting for leader for {}", ntp));
+            }
+            leader_id = instance(model::node_id{0})
+                          ->app.controller->get_partition_leaders()
+                          .local()
+                          .get_leader(ntp);
+
+            ss::sleep(100ms).get();
+        }
+        return leader_id.value();
+    }
+
+    void produce_to_partition(
+      const model::topic& topic, int partition, size_t record_count) {
+        model::ntp ntp(
+          model::kafka_namespace, topic, model::partition_id(partition));
+
+        auto leader_id = get_partition_leader(ntp);
+
+        tests::kafka_produce_transport producer(
+          instance(leader_id)->make_kafka_client().get());
+        producer.start().get();
+        auto deferred_close = ss::defer([&producer] { producer.stop().get(); });
+
+        auto records = tests::kv_t::sequence(0, record_count, partition);
+        producer
+          .produce_to_partition(topic, model::partition_id(partition), records)
+          .get();
+    }
+
+    chunked_hash_map<
+      model::topic_partition,
+      chunked_vector<model::record_batch>>
+    fetch_until_empty(direct_consumer& consumer) {
+        chunked_hash_map<
+          model::topic_partition,
+          chunked_vector<model::record_batch>>
+          ret;
+
+        while (true) {
+            auto fetched = consumer.fetch_next(1000ms).get();
+
+            if (fetched.value().empty()) {
+                break;
+            }
+            for (auto& topic_data : fetched.value()) {
+                for (auto& partition_data : topic_data.partitions) {
+                    auto& batches = ret[model::topic_partition(
+                      topic_data.topic, partition_data.partition_id)];
+
+                    std::ranges::move(
+                      partition_data.data, std::back_inserter(batches));
+                }
+            }
+        }
+        return ret;
+    }
+};
+
+TEST_F(consumer_fixture, TestBasicConsumption) {
```

Comment:
I feel like we could have more expansive tests for things like making sure this assignment epoch fencing works, etc.